spec/rollout_spec.rb

```diff
@@ -106,18 +106,39 @@
   #
   # But I think actual client is more specific
   ##
-  describe "activating a specific user" do
-    before do
-      @rollout.activate_user(:chat, double(id: 42))
+  shared_examples "User's Feature Checker" do
+    context 'when feature is rolled out for the user' do
+      subject { rollout(feature_name, user) }
```

Comment:
I don't know the common conventions in the Ruby community, so how would a reader of the contract test know that they need to implement `rollout()` and provide values for `feature_name` and `user`? Documentation? Declaring modules? Something else?

Also, i consider it perhaps a big step to both generalize the test data and the rollout action at the same time. I would probably generalize only the action and leave the test data as concrete values. One doesn't need to parameterize the data in a contract test; I only care about parameterizing the implementation of the action or the subject under test.

Reply:
Not quite sure about common convention yet, but after discussing tests for Enumerable contract I decided to go with abstract factory functions and updated it in ac591d6198858d9ec8b285bbe14db1b89056ee14 to

```ruby
  shared_examples "User's Feature Checker" do
    def rollout(feature, user)
      raise 'Subclass Responsibility'
    end

    def rollout_except(feature, user)
      raise 'Subclass Responsibility'
    end
```

perhaps name could be improved, e.g. rollout_with/without or rollout_including/excluding

Reply:
> Also, i consider it perhaps a big step to both generalize the test data

I'm agree, it also felt bit extra complex when I tried solve both at the same time.
So, I decided to revert to concrete values for now.